Add information when master node left to DiscoveryNodes' shortSummary()

tlrx opened the pull request:
This commit changes `DiscoveryNodes.Delta.shortSummary()` in order to
add information to the summary when the master node left.

Related to #28169

bleskes commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -448,7 +452,7 @@ public boolean hasChanges() {
 
         public boolean masterNodeChanged() {
-            return newMasterNode != null;
+            return newMasterNode != previousMasterNode;
```
This should use ephimeral ids, not object identity

tlrx replied:
Right, it's a bit scary, sorry.

bleskes commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -400,10 +400,14 @@ public Delta delta(DiscoveryNodes other) {
         DiscoveryNode previousMasterNode = null;
```
why not just set them both and figure things out in the delta class?

bleskes commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -478,16 +482,25 @@ public boolean added() {
-                if (newMasterNode.getId().equals(localNodeId)) {
-                    // we are the master, no nodes we removed, we are actually the first master
-                    sb.append("new_master ").append(newMasterNode());
+                if (newMasterNode() != null) {
```
do we need all this string crafting? can we just check for change and say "its was x, now it's y"?

tlrx replied:
That would simplify the summary, indeed

ywelsch replied:
Yes, there is some effort involved with string crafting. But I think it helps the end user understand what happened, for example,
- "stepped down as master" (if localNode == previousMaster && newMaster == null)
- "became master" (if localNode == newMaster)
- "lost master" (if localNode != previousMaster && newMaster == null)
- "detected master" (newMaster != null && newMaster != localNode)
- ...?

bleskes replied:
I'm good if you guys think it's worth it. My only concern is that people will start thinking stuff that are necessarily true - (or might be not true in the future) - like are we guaranteed to always have an even with a master stepped down? Maybe we have a master swap, or a restart and the intermediate cluster state was lost in batching?

tlrx replied:
We talked with @ywelsch and we agreed on logging a simple message with previous/current master values for now, as crafting more detailed messages could lead the user to expect a specific message while the related event could be lost because of cluster state batching etc. 

For now it will log messages like:
```
[node_tm0] master node changed {previous [], current [{node_tm1}{...}
[node_tm2] master node changed {previous [{node_tm2}{...}], current []}, reason: not enough master nodes
``` 
but we may revisit this in the future.

tlrx commented:
> Let's also look at the DiscoveryNodes.delta(...), which has a pretty unconventional definition of previousMasterNode and newMasterNode.

I guess this is because the Delta reports null values for previous and master nodes when it didn't change.

I changed and simplified things, please let me know if that's too much. Thanks

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -448,7 +451,9 @@ public boolean hasChanges() {
-            return newMasterNode != null;
+            String newMasterId = (newMasterNode != null) ? newMasterNode.getEphemeralId() : null;
+            String previousMasterId = (previousMasterNode != null) ? previousMasterNode.getEphemeralId() : null;
+            return Objects.equals(newMasterId, previousMasterId) == false;
```
All of this is equivalent to the simpler
```
return Objects.equals(newMasterNode, previousMasterNode) == false;
```
(see equals implementation of DiscoveryNode)

tlrx replied:
Right, thanks

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -385,26 +386,28 @@ public DiscoveryNodes newNode(DiscoveryNode node) {
+
         DiscoveryNode previousMasterNode = null;
+        if (other.masterNodeId != null) {
+            previousMasterNode = other.getMasterNode();
```
`getMasterNode()` already returns null if `masterNodeId` is null.
Maybe cleaner to make that explicit in the `getMasterNode()` method (and not rely on the map implementation to do that for us) and also use `@Nullable` on the return type. The effect is that we won't need these conditions here and can just write `return new Delta(other.getMasterNode(), getMasterNode(), ...`

tlrx commented:
I don't think it worth backporting this to 6.2.3
